docs/source/solving-interface.rst

```diff
@@ -837,8 +837,68 @@ in the transpose nullspace by providing a
 :class:`~firedrake.nullspace.VectorSpaceBasis` with the
 ``transpose_nullspace`` keyword argument to :func:`~.solve`.
 
-Singular operators in mixed spaces
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Achieving exact mean zero on nonuniform meshes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Trying the above example on a nonuniform mesh with variable cell areas
+will give a solution that does not have zero spatial mean. This is
+because PETSc uses the usual :math:`\mathbb{R}^M` inner product on the
+vectors containing the basis coefficients. In general the correct
+thing to do is to specify a basis for :math:`U \subset V'`, a
+subspace to the dual space to :math:`V`. Then the subspace we wish to
+solve the system on is :math:`\hat{V} = U^0`, the "polar space" of
+:math:`U` in :math:`V`, defined by
+
+.. math::
+
+   U^0 = \{v \in V: F[v]=0, \, \forall F \in U\}.
+
+For example, if we want to use the zero mean subspace of :math:`V`, we
+take the basis :math:`\{F_0\}`, where
+
+.. math::
+
+   F_0[v] = \int_\Omega v dx, \, \forall v \in V.
+
+To implement this, we provide a basis to PETSc consisting all of the
+assembled linear forms in the basis for :math:`U`. These are vectors
+where the i-th entry is equal to :math:`F_0[\phi_i]` where :math:`\phi_i`
+is the i-th basis vector of :math:`V``. For example, if we want to use
+:math:`F_0` above, we just use
+
+ ..  code-block:: python3
+
+     v = TestFunction(V)
+     F0 = assemble(v*dx)
+     cobasis = VectorSpaceBasis([F0])
```

Comment:
This is not correct because the basis given to PETSc is required to be orthonormal, not just orthogonal (and that orthonormality is in the R^M inner product). So you probably want to pointwise divide by:
```
with F0.dat.vec_ro as v:
    factor = v.norm()
```